File: TxtInOut/swat_u/123/SA_compute.R

rm(list = ls())
library(sensitivity)

args <- commandArgs(trailingOnly = TRUE)
wd <- args[1]
# wd <- 'G:\\SWAT\\t4\\'

setwd(wd)

basin.cnt.dat <- read.table('res_basin.dat')
hru.cnt.dat <- read.table('res_hru.dat')
output.dat <- read.table('res_output.dat')
total.output.dat <- read.table('res_total_output.dat',sep = ',')

sim.n <- length(basin.cnt.dat[,1])

params <- read.csv("params_sampling.csv",header = T)
p <- length(params[1,])

# Read simTimes n from file
settings <- file("simSetting.txt",open = "r")
Flow <- Sed <- TN <- TP <- DO <- NO3 <- NH4 <- FALSE
while (TRUE) {
  line <- readLines(settings, 1)
  if ( length(line) == 0 ) {
    break
  }
  cat(line,"\n")
  # read & create parameters
  seg <- unlist(strsplit(line, split = ":|,"))
  if(seg[1]=="simTimes"){
    n <- as.numeric(seg[2])
  }else if(seg[1]=="simTargets"){
    for(i in seg){
      if(i=="Flow"){
        Flow <- TRUE
      }else if(i=="Sed"){
        Sed <- TRUE
      }else if(i=="TN"){
        TN <- TRUE
      }else if(i=="TP"){
        TP <- TRUE
      }else if(i=="DO"){
        DO <- TRUE
      }else if(i=="NO3"){
        NO3 <- TRUE
      }else if(i=="NH4"){
        NH4 <- TRUE
      }
    }
  }
}
close(settings)
M <- 4
# n <- 65 #214 #Parameters selection for e-FAST (Saltelli, 1999, Technometrics)
omega <- numeric(p)
omega[1] <- floor((n - 1)/(2 * M))
m <- floor(omega[1]/(2 * M))
if (m >= p - 1) {
  omega[-1] <- floor(seq(from = 1, to = m, length.out = p - 1))
}else {
  omega[-1] <- (0:(p - 2))%%m + 1
}
s <- 2 * pi/n * (0:(n - 1))

n.var <- 45 ## Number of SWAT output variables 
##Define vectors to store analysis results
output <- vector('numeric', length = sim.n)
total.output <- matrix(0,  nrow= sim.n, ncol=n.var) ##all results
basin.cnt <- vector('numeric', length = sim.n) ##watershed counts
hru.cnt <- vector('numeric', length = sim.n) ##hru counts

for (i in seq(1, sim.n)) {
  basin.cnt[i] = basin.cnt.dat[i,1]
  hru.cnt[i] = hru.cnt.dat[i,1]
  output[i] = output.dat[i,1]
  # total.output[i,] = total.output.dat[i,]
}
total.output <- data.matrix(total.output.dat)
##Simulation result
fast99.result <- list(basin.cnt, hru.cnt, output, total.output)


##After simulation.
## define the result to "fast99" class to use tell and related functions
x <- list(model = NULL, M = M, s = s, omega = omega, X = params, r =fast99.result, call = match.call())
class(x) <- "fast99"

# Flow 7-6=1 <- Sed 11-6=5 <- NO3 18-6=12 <- NH4 20-6=14 <- DO 30-6=24 <- TN 48-6=42 <- TP 49-6=43

if(Flow){
  x$y <- x$r[[4]][,1] ##flow Here the code is overwritten
  ##Calculate main and interaction effects
  tell(x)
  S <- rbind(x$D1/x$V, 1 - x$Dt/x$V - x$D1/x$V)
  colnames(S) <- colnames(x$X)
  rownames(S) <- c("mainEffect", "interactions")
  write.csv(S,"SA_score_Flow.csv")
}

if(Sed){
  x$y <- x$r[[4]][,5] 
  tell(x)
  S <- rbind(x$D1/x$V, 1 - x$Dt/x$V - x$D1/x$V)
  colnames(S) <- colnames(x$X)
  rownames(S) <- c("mainEffect", "interactions")
  write.csv(S,"SA_score_Sed.csv")
}

if(NO3){
  x$y <- x$r[[4]][,12] ##NO3
  tell(x)
  S <- rbind(x$D1/x$V, 1 - x$Dt/x$V - x$D1/x$V)
  colnames(S) <- colnames(x$X)
  rownames(S) <- c("mainEffect", "interactions")
  write.csv(S,"SA_score_NO3.csv")
}


if(NH4){
  x$y <- x$r[[4]][,14] ##NH4
  tell(x)
  S <- rbind(x$D1/x$V, 1 - x$Dt/x$V - x$D1/x$V)
  colnames(S) <- colnames(x$X)
  rownames(S) <- c("mainEffect", "interactions")
  write.csv(S,"SA_score_NH4.csv")
}

if(DO){
  x$y <- x$r[[4]][,24] 
  tell(x)
  S <- rbind(x$D1/x$V, 1 - x$Dt/x$V - x$D1/x$V)
  colnames(S) <- colnames(x$X)
  rownames(S) <- c("mainEffect", "interactions")
  write.csv(S,"SA_score_DO.csv")
}

if(TN){
  x$y <- x$r[[4]][,42] 
  tell(x)
  S <- rbind(x$D1/x$V, 1 - x$Dt/x$V - x$D1/x$V)
  colnames(S) <- colnames(x$X)
  rownames(S) <- c("mainEffect", "interactions")
  write.csv(S,"SA_score_TN.csv")
}

if(TP){
  x$y <- x$r[[4]][,43] ##TP
  tell(x)
  S <- rbind(x$D1/x$V, 1 - x$Dt/x$V - x$D1/x$V)
  colnames(S) <- colnames(x$X)
  rownames(S) <- c("mainEffect", "interactions")
  write.csv(S,"SA_score_TP.csv")
}

File: TxtInOut/swat_u/123/SA_compute.py
import sys
import numpy as np
from SALib.analyze import morris
from SALib.analyze import sobol
from SALib.analyze import fast
from SALib.util import read_param_file
from SALib.test_functions import Ishigami

def compute(target):
  str_si = ""
  if method == "morris": 
    SI = morris.analyze(params, X, Y, print_to_console=True) 
    str_si = 'names,'+','.join(SI['names'])+'\n' + 'mu_star,'+','.join(str('{:.4f}'.format(i)) for i in SI['mu_star'])+'\n' + 'sigma,'+','.join(str('{:.4f}'.format(i)) for i in SI['sigma'])+'\n'
  elif method == "sobol": 
    SI = sobol.analyze(params, Y, calc_second_order=True, print_to_console=True)  
    str_si = 'names,'+','.join(params['names'])+'\n' +'S1,'+','.join(str('{:.4f}'.format(i)) for i in SI['S1'])+'\n' +'ST,'+','.join(str('{:.4f}'.format(i)) for i in SI['ST'])+'\n'
    str_si += 'S2,'
    for i in range(params['num_vars']):
      str_si += ','.join(str('{:.4f}'.format(i)) for i in SI['S2'][i])+'\n,'
  elif method == "fast":
    SI = fast.analyze(params, Y, print_to_console=True) 
    str_si = 'names,'+','.join(SI['names'])+'\n' + 'S1,'+','.join(str('{:.4f}'.format(i)) for i in SI['S1'])+'\n' + 'ST,'+','.join(str('{:.4f}'.format(i)) for i in SI['ST'])+'\n'

  with open(path+'SA_score_'+target+'.csv', 'w') as f:
    f.write(str_si)

path = sys.argv[1]

# path = "G:\\SWAT\\t_Sobol\\"
# 读取参数文件
with open(path+'paramsSum.txt', 'r') as f:
  param_list = f.readlines()

num_vars = len(param_list)
names = []
bounds = []
for i in range(0, len(param_list)):
  line = param_list[i].rstrip('\n')
  param = line.split(":")
  names.append(param[0])
  nums = param[1].split(",")
  bounds.append([float(nums[0]),float(nums[1])])

params = {
 'num_vars': num_vars,
 'names': names,
 'groups': None,
 'bounds': bounds
}

# 读取模拟方法，模拟目标
with open(path+'simSetting.txt', 'r') as f:
  sim_list = f.readlines()

method = ""
Flow = Sed = TN = TP = DO = NO3 = NH4 = False
for i in range(0, len(sim_list)):
  line = sim_list[i].rstrip('\n')
  sim = line.split(":")
  if sim[0] == "SAmethod":
    method = sim[1]
  if sim[0] == "simTargets":
    targ = sim[1].split(',')
    for j in range(0,len(targ)):
      if targ[j]=="Flow":
        Flow = True
      elif targ[j]=="Sed":
        Sed = True
      elif targ[j]=="TN":
        TN = True
      elif targ[j]=="TP":
        TP = True
      elif targ[j]=="DO":
        DO = True
      elif targ[j]=="NO3":
        NO3 = True
      elif targ[j]=="NH4":
        NH4 = True
      

# 读取采样文件
X = np.load(path+"params_sampling.npy")


# 读取结果文件
total = np.loadtxt(path+"res_total_output.dat",delimiter=",")

# 计算各个模拟目标的参数敏感性指标
if Flow:
  Y = total[:,0]
  compute("Flow")
if Sed:
  Y = total[:,4]
  compute("Sed")
if TN:
  Y = total[:,11]
  compute("TN")
if TP:
  Y = total[:,13]
  compute("TP")
if DO:
  Y = total[:,23]
  compute("DO")
if NO3:
  Y = total[:,41]
  compute("NO3")
if NH4:
  Y = total[:,42]
  compute("NH4")

